wgsl/index.bs

```diff
@@ -9586,20 +9851,26 @@ value|non-uniform=] during the execution of the function call.
 </table>
 
+To support generating [=diagnostics=], each instance of a [=CallSiteRequiredToBeUniform=] [=call site tag=] and each instance of a [=ParameterRequiredToBeUniform=]
+[=parameter tag=] also has an associated [=witness set=], which capture the root cause
+for the tags to be [=CallSiteRequiredToBeUniform=] and [=ParameterRequiredToBeUniform=], respectively.
+
 The following algorithm describes how to compute these tags for a given function:
 
 * Create nodes called <dfn noexport>RequiredToBeUniform</dfn>, <dfn noexport>MayBeNonUniform</dfn>, <dfn noexport>CF_start</dfn>, and if the function has a [=return type=] a node called <dfn noexport>Value_return</dfn>.
+    * Additionally, the [=RequiredToBeUniform=] tag has an associated [=witness set=], which is initially empty.
 * Create one node for each parameter of the function which we'll call <dfn noexport>param_i</dfn>.
 * Desugar pointers as described in [[#pointer-desugar]].
     * For each pointer parameter in the [=address spaces/function=] address space,
         create a <dfn noexport>Value_return_i</dfn> node.
-* Walk over the syntax of the function, adding nodes and edges to the graph following the rules of the next sections ([[#func-var-value-analysis]], [[#uniformity-statements]], [[#uniformity-function-calls]], [[#uniformity-expressions]]), using [=CF_start=] as the starting control-flow for the function's body.
+* Walk over the syntax of the function, adding nodes and edges to the graph, and building [=witness sets=], following the rules of the next sections ([[#func-var-value-analysis]], [[#uniformity-statements]], [[#uniformity-function-calls]], [[#uniformity-expressions]]), using [=CF_start=] as the starting control-flow for the function's body.
 * For each [=Value_return_i=] node, record which [=param_i=] nodes are reachable from it.
-* Look at which nodes are reachable from [=RequiredToBeUniform=].
-    * If this set includes the node [=MayBeNonUniform=], then reject the program.
-    * If this set includes [=CF_start=], then the [=call site tag=] for the function is [=CallSiteRequiredToBeUniform=].
+* Examine the [=RequiredToBeUniform=] node. Let *W* be its associated [=witness set=]. Let *R* be the set of nodes reachable from [=RequiredToBeUniform=].
+    * If *R* includes the node [=MayBeNonUniform=], then trigger a [=uniformity failure=].
+         * If the uniformity failure triggers [=diagnostics=], then the [=diagnostic/triggering locations=] are taken from *W*.
+    * If *R* includes [=CF_start=], then the [=call site tag=] for the function is [=CallSiteRequiredToBeUniform=], with [=witness set=] *W*.
     * Otherwise, the [=call site tag=] is [=CallSiteNoRestriction=].
-    * For each [=param_i=] in this set, the corresponding [=parameter tag=] is [=ParameterRequiredToBeUniform=].
+    * For each [=param_i=] in *R*, the corresponding [=parameter tag=] is [=ParameterRequiredToBeUniform=], with [=witness set=] *W*.
```

Comment:
This seems like it could generate diagnostics with incorrect triggering locations, which seems undesirable. Consider:
```
fn bar(c1 : bool, c2 : bool) -> vec4<f32> {
  var x : vec4<f32>;
  if c1 {
    x += textureSample(...);
  }
  if c2 {
    x + textureSample(...);
  }
  return x;
}

var a : f32;
fn foo() {
  let uniform = true;
  let nonuniform = a < 0.5;
  let x = bar(uniform, nonuniform);
}
```

It seems permissible that a diagnostic for a uniformity failure be issued with the triggering location of the first textureSample call despite that being provably uniform. This happens because both parameters in bar are reachable from RequiredToBeUniform, which has a witness set of both textureSample locations.

Reply:
Thanks for catching this again.

I've reworked so that triggering locations are tracked through the CallSiteRequiredToBeUniform and ParameterRequiredToBeUniform tags.   

Also, per community group feedback, the algorithm in the spec is much closer to a "good" implementation, as in Tint. The graph construction takes severities into account.  This lets us track (triggering location, rule) precisely according to the severity associated with the triggering call site.